### DISCUSSION TITLE
rw-ip4-in-ip6-ikev2

### DISCUSSION THREAD

**Discussion by t4n1s** [strongswan/strongswan#1969] [General]
Hi,

Running strongswan-5.9.11_2 on FreeBSD 13.2-RELEASE-p4. 

First configuration has been rw-ip4-in-ip4-ikev2 which is working like a charm. I can browse the Internet and read my mails.

So now I wanted to step up my game, so which changes are required to evolve to an rw-ip4-in-ip6-ikev2 configuration?

Unfortunately there is no example configuration in the documentation, which let's me assume this should somehow work out-of-the-box. Unfortunately it does not. I'm able to successfully establish the connection  from my client (IOS) to my VM (FreeBSD) using IPv6 and get an IPv4 address from the rw-ikev2-pool assigned, but I'm unable to reach the client's IPv4 address from the other side of the tunnel. Is there some kind of internal routing which needs to be set in strongswan so I can force packets to enter the tunnel ? 

Any advice would be appreciated.

Thank you in advance.

**Comment by tobiasbrunner** [comment]
> Is there some kind of internal routing which needs to be set in strongswan so I can force packets to enter the tunnel ?

Depends on the host from which you send traffic. On FreeBSD, when sending from the gateway, you might have to enable `charon.plugins.kernel-pfkey.route_via_internal`. But I guess that would be the same for IPv4. Could always be an issue with FreeBSD's IPsec stack.

**Comment by t4n1s** [comment]
Thx for the reply. Just for the record all this is experienced by utilizing vici instead of stroke.

The mentioned flag [charon.plugins.kernel-pfkey.route_via_internal](https://docs.strongswan.org/docs/5.9/config/strongswanConf.html#_charon_plugins_kernel_pfkey) has not been set, so it's still running in default setting.

I followed up on your advice and tested the exact same configuration on Alma 9. To make it work I needed to change `local_ts=192.168.10.254{tcp/993],192.168.10.254[tcp/587]` to `local_ts=0.0.0.0/0`. Another point I noticed on Alma 9 an ipsec0 interface gets created as soon as StrongSwan has been launched, but no such interface get's created on FreeBSD 13. Reading the documentation I have been under the impression that an interface is not needed as Policy-based Routing is the default.  The last thing I noticed for whatever reason after a few connection under FreeBSD I'm unable to connect to StrongSwan using IPv6 any further and the following error message appears: `charon[xxxx] 04[NET] error writing to socket: No route to host`. Rebooting the FreeBSD VM helps to establish a connection again, but after some time it just stops working and I have to reboot again.

Thank you for your help in advance.

**Reply by tobiasbrunner** [reply]
> Another point I noticed on Alma 9 an ipsec0 interface gets created as soon as StrongSwan has been launched, but no such interface get's created on FreeBSD 13.

The package or build you are using there seems to have the [_kernel-libipsec_ plugin](https://docs.strongswan.org/docs/5.9/plugins/kernel-libipsec.html) enabled. Since that doesn't support protocol/ports in policies that also explains why you had to change the traffic selector.

**Reply by t4n1s** [reply]
`pkg info strongswan | grep KERNELLIBIPSEC`
 `	KERNELLIBIPSEC : off`

Nope. =)

According to strongSwan documentation, that option is off by default and is also not recommended.

**Reply by tobiasbrunner** [reply]
Nope what? I was referring to the Alma 9 package, with which that device was created, not the FreeBSD package.

**Comment by t4n1s** [comment]
Update:

Looks like there is an issue with installing the route into the FreeBSD kernel, I get the following error message :
`charon[2172]: 09[KNL] adding PF_ROUTE route failed: Invalid argument`
`charon[2172]: 09[KNL] installing route failed: 10.0.10.1/32 via fe80::XXXX:XXX:XXXX:XXXX src %any dev vtnet0`
`charon[2172]: 09[IKE] CHILD_SA RoadWarrior-ikev2-mschap{1} established with SPIs cb322918_i 02e04d80_o and TS 0.0.0.0/0 === 10.0.10.1/32`

Unfortunately I'm unable to extract anymore information by increasing the loglevel for the subsystems ike and knl.

**Reply by tobiasbrunner** [reply]
Might be the same issue that was noticed with IPv6 routes and virtual IPs on FreeBSD (see [this old and still unresolved ticket](https://wiki.strongswan.org/issues/3285)).

**Reply by t4n1s** [reply]
Thx for the reply and that ticket. What strikes me as a bit odd, is the fact, that if I establish the connection using IPv4 by utilizing the exact same configuration, all works perfectly fine.

**Comment by t4n1s** [comment]
Looks like this is related to the rewrite in routing in FreeBSD [Bug 255678 - security/strongswan cant add routes via RTM_ADD via PF_ROUTE socket](https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=255678)

**Reply by t4n1s** [reply]
I was wrong.

Disabling MOBIKE makes the RW 4in6 scenario work on FreeBSD.
